lib/utils/memoization.jl

```diff
@@ -40,14 +40,17 @@ macro memoize(ex...)
     # anything, that entry will be the memoized new_value, or else a dictionary of values.
     @gensym global_cache
 
+    # in the presence of thread adoption, we need to use the maximum thread ID
+    nthreads = :( VERSION >= v"1.9" ? Threads.maxthreadid() : Threads.nthreads() )
+
     # generate code to access memoized values
     # (assuming the global_cache can be indexed with the thread ID)
     if key === nothing
         # if we don't have to key on anything, use the global cache directly
         global_cache_eltyp = :(Union{Nothing,$rettyp})
         ex = quote
             cache = get!($(esc(global_cache))) do
-                $global_cache_eltyp[nothing for _ in 1:Threads.nthreads()]
+                $global_cache_eltyp[nothing for _ in 1:$nthreads]
```

Comment:
We might need to resize `cache` when maxthreadid changes?

Reply:
That's being done, although in a roundabout way: The LazyInitialized takes a `validator` argument now, and that's used to trigger re-initialization: https://github.com/JuliaGPU/CUDA.jl/blob/d4eb8d08a37111227cb29d000f2c1bce5fa1d234/lib/utils/memoization.jl#L113

I was very close to ripping it all out and replacing it with a simple lock (also because there's a race now between the value being returned from `get!` and the actual use), but I guess it was worth it when I first developed this...